File: harvester/src/main/java/harvester/KeywordHarvester.java
package harvester;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.commons.collections15.Transformer;

public class KeywordHarvester {
    private final static RelatedKeyword[] EMPTY_RESULT = new RelatedKeyword[0];

    private PageLoader pageLoader;

    private final Map<String, Integer> wordCount = new HashMap<String, Integer>();
    private String currentKeyword;

    private static final Set<String> STOP_WORDS = new HashSet<String>(Arrays.asList("and", "of", "for", "with", "or"));

    private static final String SPLIT_PATTERN = "[/, :;]";

    public synchronized RelatedKeyword[] findRelatedKeywords(String word) {
	if (word == null || "".equals(word.trim())) {
	    return EMPTY_RESULT;
	}

	try {
	    this.currentKeyword = word;
	    collectRelatedKeywords(pageLoader.getPagesForKeyword(word));

	    return createResult();
	} finally {
	    this.currentKeyword = null;
	    wordCount.clear();
	}
    }

    private RelatedKeyword[] createResult() {
	List<RelatedKeyword> result = new ArrayList<RelatedKeyword>();
	for (Map.Entry<String, Integer> entry : wordCount.entrySet()) {
	    result.add(new RelatedKeyword(entry.getKey(), entry.getValue()));
	}

	Collections.sort(result);
	return result.toArray(new RelatedKeyword[result.size()]);
    }

    private void collectRelatedKeywords(Iterable<HtmlPageData> pages) {
	for (HtmlPageData page : pages) {
	    collect(page, 5, new Transformer<HtmlPageData, String[]>() {
		public String[] transform(HtmlPageData input) {
		    return input.getTitle().split(SPLIT_PATTERN);
		}
	    });
	    collect(page, 2, new Transformer<HtmlPageData, String[]>() {
		public String[] transform(HtmlPageData input) {
		    return input.getMetaValue().split(SPLIT_PATTERN);
		}
	    });
	    collectBodyWords(page, 1);
	}
    }

    private void collectBodyWords(HtmlPageData source, int weight) {
        for (final ElementQualifier qualifier : ElementQualifier.values()) {
            collect(source, weight, new Transformer<HtmlPageData, String[]>() {
                public String[] transform(HtmlPageData input) {
                    return input.getBodyValue(qualifier).split(SPLIT_PATTERN);
                }
            });
        }
    }

    private void collect(HtmlPageData page, int weight, Transformer<HtmlPageData, String[]> transformer) {
	for (String string : transformer.transform(page)) {
	    String trimmedString = string.trim().toLowerCase();
	    if (isStringValid(trimmedString)) {
		Integer previous = wordCount.remove(trimmedString);
		if (previous == null) {
		    wordCount.put(trimmedString, weight);
		} else {
		    previous += weight;
		    wordCount.put(trimmedString, previous);
		}
	    }
	}
    }

    private boolean isStringValid(String trimmedString) {
	if (trimmedString == null || 0 == trimmedString.length()) {
	    return false;
	}

	if (currentKeyword.equals(trimmedString)) {
	    return false;
	}

	if (STOP_WORDS.contains(trimmedString)) {
	    return false;
	}

	if (trimmedString.length() < 3) {
	    return false;
	}

	if (isDigit(trimmedString)) {
	    return false;
	}

	return true;
    }

    private boolean isDigit(String sting) {
	return sting.matches("\\d*\\.\\d+");
    }

    public void setPageLoader(PageLoader pageLoader) {
	this.pageLoader = pageLoader;
    }
}

File: harvester-scala/src/main/scala/harvester/keywordharvester.scala
package harvester

class HarvesterException(message: String, t: Throwable) extends RuntimeException(message, t)

case class RelatedKeyword(word: String, relevance: Int) extends Comparable[RelatedKeyword] {
  def compareTo(o: RelatedKeyword): Int = {
    return o.relevance - this.relevance
  }
}

object ElementQualifier extends Enumeration {
  type ElementQualifier = Value
  val H1, H2, H3 = Value

  def valueOfTag(value: String) = H1
}

import harvester.ElementQualifier._

class HtmlPageData(val title: String, val metaValue: String, val bodyContent: Map[ElementQualifier, String]) {
  def getBodyValue(qualifier: ElementQualifier) = bodyContent.get(qualifier).getOrElse("")
  def getTitle = ""
  def getMetaValue = ""

}

object KeywordHarvester {
  val EMPTY_RESULT = Array.empty[RelatedKeyword]
  val STOP_WORDS = Set("and", "of", "for", "with", "or")
  val SPLIT_PATTERN = "[/, :;]";

}

import KeywordHarvester._

class KeywordHarvester(private val keyword: String) {
  var pageLoader: PageLoader = _
  val wordCount = scala.collection.mutable.Map.empty[String, Int]

  def findRelatedKeywords(): List[RelatedKeyword] = {
    if (keyword == null || ("" == keyword.trim)) {
      return List.empty
    }
    collectRelatedKeywords(pageLoader.getPagesForKeyword(keyword))
    wordCount.toSet.map({ entry: Pair[String, Int] => RelatedKeyword(entry._1, entry._2) }).toList
  }

  private def collectRelatedKeywords(pages: Iterable[HtmlPageData]): Unit = {
    for (page <- pages) {
      collect(page, 5, _.title)
      collect(page, 2, _.metaValue)
      ElementQualifier.values.foreach({ q: ElementQualifier => collect(page, 1, _.getBodyValue(q)) })
    }
  }

  private def collect(page: HtmlPageData, weight: Int, transformer: (HtmlPageData => String)): Unit = {
    for (string <- transformer(page).split(SPLIT_PATTERN)) {
      val trimmedString = string.trim.toLowerCase
      if (isStringValid(trimmedString)) {
        wordCount.remove(trimmedString) match {
          case Some(value) => wordCount.put(trimmedString, value + weight)
          case None => wordCount.put(trimmedString, weight)
        }
      }
    }
  }

  private def isStringValid(trimmedString: String): Boolean = {
    if (trimmedString == null || 0 == trimmedString.length) {
      return false
    }
    if (keyword == trimmedString) {
      return false
    }
    if (STOP_WORDS.contains(trimmedString)) {
      return false
    }
    if (trimmedString.length < 3) {
      return false
    }
    if (isDigit(trimmedString)) {
      return false
    }
    return true
  }

  private def isDigit(sting: String): Boolean = {
    return sting.matches("\\d*\\.?\\d+")
  }
}